Fix 9307 - raise self (Failure) in python2, wrapped exception in python3

jwtrhs opened the pull request:
Trac bug reference: https://twistedmatrix.com/trac/ticket/9307

Looks like this was last working in 16.6.0, but became broken in 17+.

jwtrhs committed:
Fix 9307 - raise self (Failure) in python2, wrapped exception in python3

jwtrhs committed:
Add news fragment

jwtrhs committed:
Edit news fragment

exarkun reviewed (dismissed):
Thanks.  Apparently this behavior has no automated test coverage (if you can believe that).  Can you please add a couple unit tests as well?  Thanks again.

jwtrhs committed:
Add test cases for Failure.trap

jwtrhs committed:
Fix test case descriptions

jwtrhs commented:
No worries, I've added a couple test cases around expected trap() behaviour.

jwtrhs committed:
Improve existing tests and remove duplicates test

jwtrhs commented:
Actually turns out there were already some tests in src/twisted/test/test_failure.py but they only covered the Python3 expected functionality. Modified the test to only run on Python3 and added a new test for Python2.

markrwilliams commented on src/twisted/test/test_failure.py:
```diff
@@ -65,6 +65,9 @@ def test_trapRaisesWrappedException(self):
         C{Exception}.
         """
+        if not _PY3:
+            raise SkipTest('Only relevant on Python 3.')
```
It would be nice if this message were more informative.  Twisted raises the wrapped exception in Python 3 because before [5519](https://github.com/twisted/twisted/pull/843/files), `Failure` did not inherit from `BaseException`, and Python 3 can only `raise` instances of `BaseException`.  Now that `Failure` _is_ a `BaseException`, this behavior persists because it's the API, and as this PR demonstrates, it's bad to break that API.  Even just a `@see` directed at [the documentation](http://twisted.readthedocs.io/en/latest/core/howto/python3.html#twisted-python-failure) would be great.

markrwilliams commented on src/twisted/test/test_failure.py:
```diff
@@ -75,6 +78,24 @@ def test_trapRaisesWrappedException(self):
+        expected types, L{failure.Failure.trap} raises itself.
+        """
+        if _PY3:
+            raise SkipTest("Only relevant on Python 2.")
```
See above - Python 2 has raised the `Failure` instance for a long time.

jwtrhs committed:
Add link to API differences

markrwilliams commented:
@jwaterhouse Looks like twistedchecker doesn't like the format of the comments: https://travis-ci.org/twisted/twisted/jobs/288064570#L537

jwtrhs committed:
Add new line between description and tag

jwtrhs committed:
Reformat skip test comments.

jwtrhs committed:
Wrap skip test explanantion in triple quotes to prevent tripping checker

jwtrhs commented:
Took a couple tries, but finally figured out the comment checker was picking up the '#' in the @see URL as a comment and emitting a couple errors. But have put in a workaround and flagged the issue here:

https://github.com/twisted/twistedchecker/issues/135

markrwilliams commented:
@jwaterhouse This is fine for now.  Thanks so much for your contribution!
